### ISSUE TITLE
Significant latency under macOS on a Managed machine

### ISSUE THREAD

**Issue by agucova** [pyenv/pyenv#2375]
I'm getting a pretty significant slowdown on all pyenv commands, even getting the version or the help prompt:
```shell
➜ time pyenv --version
pyenv 2.3.0

Executed in  968.77 millis    fish           external
   usr time    8.76 millis    0.12 millis    8.64 millis
   sys time   26.50 millis    2.27 millis   24.23 millis
````

```shell
➜ time pyenv --help
Usage: pyenv <command> [<args>]

...

Executed in    5.81 secs      fish           external
   usr time  121.37 millis    0.15 millis  121.22 millis
   sys time  267.39 millis    1.75 millis  265.64 millis
```

This seems to happen regardless of the shell being used (fish, zsh, bash or sh), the installation method (homebrew or automatic installer) and whether there are any python versions installed. The program does work correctly, but the latency has become too much for simple operations. This is a newly bought last gen MacBook Pro 16" with no general performance issues. 

The latency does not seem to be isolated to the startup of the command either, by example, there is latency on the appearance of each successive command in the help prompt.

What is even weirder is this seems to affect upstream `rbenv` too, and I can't seem to isolate the slowdown to bash itself or to any particular system call after system tracing. I'm at a loss as to how to keep debugging this, and I can't rule that this might be a more general issue with the Kernel, SIP, or something else.

### Prerequisite
* [X] Make sure your problem is not listed in [the common build problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
* [X] Make sure no duplicated issue has already been reported in [the pyenv issues](https://github.com/pyenv/pyenv/issues). You should look for closed issues, too.
* [X] Make sure you are not asking us to help solving your specific issue.
* [X] Make sure your problem is not derived from packaging (e.g. [Homebrew](https://brew.sh)).
* [X] Make sure your problem is not derived from plugins.

### Description
- [X] Platform information (e.g. Ubuntu Linux 16.04): macOS Monterey 12.4
- [X] OS architecture (e.g. amd64): arm64 (Apple Silicon)
- [X] pyenv version: 2.3.0, also affecting rbenv 1.2.0
- [X] Python version: All versions
- [X] C Compiler information (e.g. gcc 7.3): gcc 13.1.6
- [X] Please attach the debug trace of the failing command as a gist: [`trace.log`](https://gist.github.com/agucova/7fc9d8c3155cf0c11fbec4baba874216)

**Comment by native-api** [comment]
WFM on a Hackintosh VM:

```
% \time -h pyenv --version
pyenv 2.3.0-11-ge676fde9
	0,19s real		0,03s user		0,07s sys
admin@Mac-Admin .pyenv % \time -h pyenv --help   
Usage: pyenv <command> [<args>]
<...>
	1,60s real		0,22s user		0,51s sys
%
```

**Comment by native-api** [comment]
My blind guesses are an antivirus, including Apple's built-in one (other users have already reported it disrupting Pyenv operation), or your Bash runs in Rosetta.

**Comment by agucova** [comment]
I can clarify that bash is not running on Rosetta (it's the default binary shipped with macOS, which is arm64).

As for Apple's Gatekeeper or XProtect, I can't seem to find any obvious signs of them acting on the system profile I took. While I can't rule them out, I did notice most of the time is spent waiting on threads:

![image](https://user-images.githubusercontent.com/4694408/169620060-15d19500-d5e9-4f6f-9ee9-ba7c6dd44ae7.png)

![image](https://user-images.githubusercontent.com/4694408/169620331-e1073488-1579-4394-87ea-f124128f440b.png)

**Comment by native-api** [comment]
> I did notice most of the time is apparently spent waiting on threads

That's not too surprising. It's probably waiting for subprocesses -- Bash code is very fork-heavy.

Okay, let's profile the code and see if there are any obvious hogs.

Make sure you have the `coreutils` Homebrew package installed.
Then add `$(gdate "+%s.%N")` into the `PS4` value in the Pyenv executable near the beginning (as per https://stackoverflow.com/questions/5014823/how-can-i-profile-a-bash-shell-script-slow-startup) and run it with tracing again.

**Comment by giufus** [comment]
Could this coming from redundant setup instructions in .zshrc ? I see that every time the shell is loaded the following lines are appended to my $HOME/.zshrc 

```
export PYENV_ROOT="$HOME/.pyenv"
command -v pyenv >/dev/null || export PATH="$PYENV_ROOT/bin:$PATH"
eval "$(pyenv init -)"
```

**Comment by agucova** [comment]
@giufus given that this happens in any shell, that doesn't seem plausible.

**Comment by hassec** [comment]
@agucova did you ever find a solution to this problem?

**Comment by agucova** [comment]
> @agucova did you ever find a solution to this problem?

I didn't. I spent far too many hours trying to figure out what happened (10+), so I gave up and ended up using Homebrew to manage my python versions (I know, cursed).

**Comment by hassec** [comment]
Thanks for the quick reply @agucova, I guess I'll just have to do without pyenv for now then :/. 

I've already tried a bunch of profiling but am a bit at a loss at what it could be. @native-api if there are any debugging step or similar that could help please let me know

**Comment by native-api** [comment]
> @native-api if there are any debugging step or similar that could help please let me know



> Make sure you have the `coreutils` Homebrew package installed.
> Then add `$(gdate "+%s.%N")` into the `PS4` value in the Pyenv executable near the beginning (as per https://stackoverflow.com/questions/5014823/how-can-i-profile-a-bash-shell-script-slow-startup) and run it with tracing again.

If you publish the trace of a problematic command with the above addition, we can try to figure out what takes it so long.

**Comment by native-api** [comment]
> Then add `$(gdate "+%s.%N")` into the `PS4` value in the Pyenv executable near the beginning

Just to be clear, by the above, I mean prepending "`$(gdate "+%s.%N")`" to the PS4 value at https://github.com/pyenv/pyenv/blob/8b0c743f122a29023797047f9b5fe4c833fb5902/libexec/pyenv#L11 .

**Comment by hassec** [comment]
@native-api if I prepend that bit to `PS4` it seems that the script is somehow aborted, see short output below... 🤔 But I've not yet dug deeper because I'm wondering if this might not be a pyenv problem... 

After digging some more I noticed that other bash based applications (e.g. brew) are also slow and running 
```
time for i in $(seq 1 100); do bash --norc -i -c exit; done
```
takes up to ~8seconds, while it only takes ~0.1s or less on all my other machines. Since this is a work laptop I'm now a bit suspicious if there is maybe some endpoint security or similar software running that might just in general make launching new things very slow. And I guess because pyenv launches a large amount of bash processes it might just be impacted more severely.




```
1695158258.953501000 (/Users/chasse/software/bin/pyenv:23): enable -f /Users/chasse/software/bin/../libexec/pyenv-realpath.dylib realpath
1695158258.965086000 (/Users/chasse/software/bin/pyenv:30): '[' -z '' ']'
11695158258.975416000 (/Users/chasse/software/bin/pyenv:32): type -P readlink
1695158259.031850000 (/Users/chasse/software/bin/pyenv:32): READLINK=/usr/bin/readlink
1695158259.037394000 (/Users/chasse/software/bin/pyenv:33): '[' -n /usr/bin/readlink ']'
1695158259.043600000 (/Users/chasse/software/bin/pyenv:58): '[' -z /Users/chasse/software/pyenv ']'
1695158259.050728000 (/Users/chasse/software/bin/pyenv:61): PYENV_ROOT=/Users/chasse/software/pyenv
```

**Comment by native-api** [comment]
> it seems that the script is somehow aborted



> there is maybe some endpoint security or similar software running that might just in general make launching new things very slow

This reminds me of exactly this problem recently reported, exactly with MacOS' parental control software. Can't find it now, but the problem was exactly that it queried the 'launch control' daemon at every process start and it couldn't cope, eventually erroed on a timeout and the OS killed the build.

**Comment by native-api** [comment]
> Since this is a work laptop I'm now a bit suspicious if there is maybe some endpoint security or similar software running that might just in general make launching new things very slow.

Looks like https://github.com/pyenv/pyenv/issues/2588 .

**Comment by nexgen** [comment]
> Could this coming from redundant setup instructions in .zshrc ? I see that every time the shell is loaded the following lines are appended to my $HOME/.zshrc
> 
> ```
> export PYENV_ROOT="$HOME/.pyenv"
> command -v pyenv >/dev/null || export PATH="$PYENV_ROOT/bin:$PATH"
> eval "$(pyenv init -)"
> ```

Did you find any solution. I'm having the exact same problem.

**Comment by native-api** [comment]
Closing as duplicate of https://github.com/pyenv/pyenv/issues/2588
